### PR TITLE
New Polynomial Filtering System

### PR DESCRIPTION
**Pull request by EthanBullVulpe**
This pull request brings in a new polynomial filtering system that can be used to filter terms out of a polynomial.

filters can be specified in the poly.in file as such:

add_filter['type', 'arg1', 'arg2', ...]

There are currently 2 types of filters: 'degree' and 'not'.

The degree filter filters monomials based on the degree of variables in the monomial
The not filter simply inverts another type of filter.

Here is the syntax for the degree filter:
add_filter['degree', 'variable_string', 'degree_string', 'term_string']
Here are some examples of what you can do with it:
* add_filter['degree', 'x-AA', '2+', '4/5'] -- filters out all monomials with total degree = 4 or 5 and degree in a x-AA variable >= 2
* add_filter['degree', 'x-intra-AB/x-BB', '2-', '2-4'] -- filters out all monomials with total degree in the inclusive range 2-4 and degree in a x-intra-AB or x-BB variable of <= 2 (note that this includes terms of degree 0 in these variables)
* add_filter['degree', 'x-intra-**', '1+', '*'] -- filters out all monomials with ANY total degree and degree >= 1 in any intramolecular variable.

And a lot more, see the constructor comment for DegreeFilter for more information.

In order to use the not filter, just add an additional 'not' argument at the beggining, so:
add_filter['not', 'degree', ''x-*-**', '3+', '*'] - filter out all terms that do NOT have at least 1 variable with degree greater than or equal to 3.

While not all filtering options are implemented, I think this is a good place to start, as we can filter a lot of things with just these two filters.

### CODE REVIEW

File: polynomial_generation/src/filters.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                        * y+        -- affects monomials with TOTAL degree equal to or greater than y
+                        * y-z       -- affects monomials with TOTAL degree in the range [y, z] (inclusive)
+                        * y         -- affects monomials with TOTAL degree y
+                        * *         -- affects monomials regardless of their TOTAL degrees
+                    y and z can be any integer >= 0
+
+        Returns:
+            A new DegreeFilter
+
+        """
```
[the remark this answers is not in the document]
Comment (agoetz) [reply polynomial_generation/src/filters.py]: Yes, this seems clear to me.

[the remark this answers is not in the document]
Comment (agoetz) [reply polynomial_generation/src/filters.py]: We should add a note that the implementation works only for degrees < 10. I don't think we will ever want that high degrees but it should be documented.

File: polynomial_generation/src/generate_input_poly.py [hunk 2]
```diff
@@ -140,6 +140,26 @@ def generate_input_poly(settings_file, poly_input):
                         combined_letters = ''.join(sorted(split_atom1[0] + split_atom2[0]))
                         poly_in_file.write("add_variable['" + split_atom1[0] + split_atom1[1] + "', '" + split_atom1[2] + "', '" + split_atom2[0] + split_atom2[1] + "', '" + split_atom2[2] + "', 'x-" + combined_letters + "']\n")
 
+        # newline between variables and filters
+        poly_in_file.write("\n")
+
+        # add filter based on the filtering setting in settings.ini
+        polynomial_filtering = settings.get("poly_generation", "accepted_terms", "all" if len(fragments) == 1 else "purely-inter")
```
Comment (agoetz) [inline polynomial_generation/src/generate_input_poly.py]: This implies default filtering, do we issue a warning to the user somewhere?

**Comment by agoetz** [comment]
Very nice! The filtering code is well written. We should test it on the original MB-pol polynomials to confirm that we can re-generate the published polynomials using your new polynomial generation code and filters.

File: polynomial_generation/src/filters.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EthanBullVulpe) [reply polynomial_generation/src/filters.py]: This is true, also the polynomial generator and fitcode will break if there are more than 10 atoms of 1 type in a monomer.

File: polynomial_generation/src/generate_input_poly.py (hunk 2 above)
Comment (EthanBullVulpe) [reply polynomial_generation/src/generate_input_poly.py]: yes, we should. adding a notification when a default value is used is a good idea. I think this should go in the acessor methods of the SettingsReader class.

**Pull request merged**